src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -371,146 +371,154 @@ internal unsafe void Start()
                 if (context is null)
                 {
                     // Flow suppressed, just run here
-                    ProcessEvents(numEvents.ToInt32(), eventPaths, eventFlags, eventIds, watcher);
+                    ProcessEvents(numEvents.ToInt32(), eventPaths, new Span<FSEventStreamEventFlags>(eventFlags, numEvents.ToInt32()), new Span<FSEventStreamEventId>(eventIds, numEvents.ToInt32()), watcher);
                 }
                 else
                 {
                     ExecutionContext.Run(
                         context,
-                        (object o) => ((RunningInstance)o).ProcessEvents(numEvents.ToInt32(), eventPaths, eventFlags, eventIds, watcher),
+                        (object o) => ((RunningInstance)o).ProcessEvents(numEvents.ToInt32(), eventPaths, new Span<FSEventStreamEventFlags>(eventFlags, numEvents.ToInt32()), new Span<FSEventStreamEventId>(eventIds, numEvents.ToInt32()), watcher),
                         this);
                 }
             }
 
             private unsafe void ProcessEvents(int numEvents,
                 byte** eventPaths,
-                FSEventStreamEventFlags* eventFlags,
-                FSEventStreamEventId* eventIds,
+                Span<FSEventStreamEventFlags> eventFlags,
+                Span<FSEventStreamEventId> eventIds,
                 FileSystemWatcher watcher)
             {
-                // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
-                // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
-                List<FSEventStreamEventId> handledRenameEvents = null;
-                Memory<char>[] events = new Memory<char>[numEvents];
-                ParseEvents();
+                // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll store the event id
+                // so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
+                int? handledRenameEvents = null;
```

Comment:
Why is it ok to switch from the list to the single value?  IIRC, it was possible that the two events for a rename weren't immediately next to each other.  Is that not the case?  They're guaranteed to be adjacent?

Reply:
The list of handled events didn't make sense even before as the matching event was the first rename event following the current event. Then this rename event was skipped when processed (not removed from the list) and then the same happened again. The list growed however only last event was considered to skip.